openmc/arithmetic.py

```diff
@@ -0,0 +1,842 @@
+        self._aggregate_op = None
+
+        if nuclides is not None:
+            self.nuclides = nuclides
+        if aggregate_op is not None:
+            self.aggregate_op = aggregate_op
+
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __eq__(self, other):
+        return str(other) == str(self)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __repr__(self):
+
+        # Append each nuclide in the aggregate to the string
+        string = '{0}('.format(self.aggregate_op)
+        names = [nuclide.name if isinstance(nuclide, openmc.Nuclide)
+                 else str(nuclide) for nuclide in self.nuclides]
+        string += ', '.join(map(str, names)) + ')'
+        return string
+
+    @property
+    def nuclides(self):
+        return self._nuclides
+
+    @property
+    def aggregate_op(self):
+        return self._aggregate_op
+
+    @property
+    def name(self):
+
+        # Append each nuclide in the aggregate to the string
+        names = [nuclide.name if isinstance(nuclide, openmc.Nuclide)
+                 else str(nuclide) for nuclide in self.nuclides]
+        string = '(' + ', '.join(map(str, names)) + ')'
+        return string
+
+    @nuclides.setter
+    def nuclides(self, nuclides):
+        cv.check_iterable_type('nuclides', nuclides, (str, CrossNuclide))
+        self._nuclides = nuclides
+
+    @aggregate_op.setter
+    def aggregate_op(self, aggregate_op):
+        cv.check_type('aggregate_op', aggregate_op, str)
+        cv.check_value('aggregate_op', aggregate_op, _TALLY_AGGREGATE_OPS)
+        self._aggregate_op = aggregate_op
+
+
+class AggregateFilter(object):
+    """A special-purpose tally filter used to encapsulate an aggregate of a
+    subset or all of a tally filter's bins for tally aggregation.
+
+    Parameters
+    ----------
+    aggregate_filter : Filter or CrossFilter
+        The filter included in the aggregation
+    bins : Iterable of tuple
+        The filter bins included in the aggregation
+    aggregate_op : str
+        The tally aggregation operator (e.g., 'sum', 'avg', etc.) used
+        to aggregate across a tally filter's bins with this AggregateFilter
+
+    Attributes
+    ----------
+    type : str
+        The type of the aggregatefilter (e.g., 'sum(energy)', 'sum(cell)')
+    aggregate_filter : filter
+        The filter included in the aggregation
+    aggregate_op : str
+        The tally aggregation operator (e.g., 'sum', 'avg', etc.) used
+        to aggregate across a tally filter's bins with this AggregateFilter
+    bins : Iterable of tuple
+        The filter bins included in the aggregation
+    num_bins : Integral
+        The number of filter bins (always 1 if aggregate_filter is defined)
+
+    """
+
+    def __init__(self, aggregate_filter=None, bins=None, aggregate_op=None):
+
+        self._type = '{0}({1})'.format(aggregate_op,
+                                       aggregate_filter.short_name.lower())
```

Comment:
again, the `aggregate_filter` parameter is defaulted to None but its attributes are accessed immediately which would result in an AttributeError. Is there a specific reason to default all these parameters?